### PR TITLE
Add support for the TLS registry to OIDC and OIDC client extensions

### LINKED ISSUE #41001: Add support for the TLS registry to the OIDC Common HTTP client
### Description

With the integrated TLS registry, it should be possible to configure TLS using the TLS registry instead of the specific OIDC common configuration.

### Implementation ideas

This is the code used for the mailer:

```java
 private void configureTLS(String name, MailerRuntimeConfig config, TlsConfigurationRegistry tlsRegistry, MailConfig cfg,
            boolean globalTrustAll) {
        TlsConfiguration configuration = null;

        // Check if we have a named TLS configuration or a default configuration:
        if (config.tlsConfigurationName.isPresent()) {
            Optional<TlsConfiguration> maybeConfiguration = tlsRegistry.get(config.tlsConfigurationName.get());
            if (!maybeConfiguration.isPresent()) {
                throw new IllegalStateException("Unable to find the TLS configuration "
                        + config.tlsConfigurationName.get() + " for the mailer " + name + ".");
            }
            configuration = maybeConfiguration.get();
        } else if (tlsRegistry.getDefault().isPresent() && tlsRegistry.getDefault().get().isTlsEnabled()) {
            configuration = tlsRegistry.getDefault().get();
        }

       // Apply the configuration
        if (configuration != null) {
            // This part is often the same (or close) for every Vert.x client:
            cfg.setSsl(true);

            if (configuration.getTrustStoreOptions() != null) {
                cfg.setTrustOptions(configuration.getTrustStoreOptions());
            }

           // For mTLS:
            if (configuration.getKeyStoreOptions() != null) {
                cfg.setKeyCertOptions(configuration.getKeyStoreOptions());
            }

            if (configuration.isTrustAll()) {
                cfg.setTrustAll(true);
            }
            if (configuration.getHostnameVerificationAlgorithm().isPresent()) {
               // ACHTUNG HERE - this is protocol specific. The HTTP-based protocols should use HTTPS by default. 
                cfg.setHostnameVerificationAlgorithm(configuration.getHostnameVerificationAlgorithm().get());
            }

            SSLOptions sslOptions = configuration.getSSLOptions();
            if (sslOptions != null) {
                cfg.setSslHandshakeTimeout(sslOptions.getSslHandshakeTimeout());
                cfg.setSslHandshakeTimeoutUnit(sslOptions.getSslHandshakeTimeoutUnit());
                for (String suite : sslOptions.getEnabledCipherSuites()) {
                    cfg.addEnabledCipherSuite(suite);
                }
                for (Buffer buffer : sslOptions.getCrlValues()) {
                    cfg.addCrlValue(buffer);
                }
                cfg.setEnabledSecureTransportProtocols(sslOptions.getEnabledSecureTransportProtocols());

            }

        } else {
           // Mailer specific configuration (very incomplete as you can see:
            boolean trustAll = config.trustAll.isPresent() ? config.trustAll.get() : globalTrustAll;
            cfg.setSsl(config.ssl);
            cfg.setTrustAll(trustAll);
            applyTruststore(config, cfg);
        }
    }
```

### PR DESCRIPTION
**Pull request by michalvavrik**
- Closes: https://github.com/quarkusio/quarkus/issues/41001

### CODE REVIEW

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 1]
```diff
@@ -197,8 +197,9 @@ quarkus.oidc.credentials.jwt.key-file=privateKey.pem
 ----
 quarkus.oidc.auth-server-url=http://localhost:8180/realms/quarkus/
 quarkus.oidc.client-id=quarkus-app
-quarkus.oidc.credentials.jwt.key-store-file=keystore.jks
-quarkus.oidc.credentials.jwt.key-store-password=mypassword
+quarkus.oidc-client.credentials.jwt.tls-configuration-name=oidc-client
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: In this case it looks like an extra property to set without any advantage. IMHO having an option to refer to the specific TLS configuration is great, but it should not be a prerequisite if there is no intention to have a given TLS configuration shared

**Comment by sberyozkin** [comment]
Hi Michal, @michalvavrik Thanks for this PR, and I'm sure it will work out well, but my immediate concern is that I'm seeing OIDC users are asked to enter one more configuration property and I don't see what is the point. So before me going further with the review, I'd like us agree on it. 
As far as I understand, referring to a named TLS configuration should be optional. I can imagine where the same TLS configuration is shared between different Quarkus systems, so referring to it would be nice, but right now, I'm not seeing why we should update all properties to use a named TLS configuration

**Comment by michalvavrik** [comment]
> Hi Michal, @michalvavrik Thanks for this PR, and I'm sure it will work out well, but my immediate concern is that I'm seeing OIDC users are asked to enter one more configuration property and I don't see what is the point. So before me going further with the review, I'd like us agree on it.

Old code still works, I simply c&p original tests and changed config properties, original tests are still in place. Dropping deprecation of config properties and of docs adjustments is matter of minute. I can do that.

> As far as I understand, referring to a named TLS configuration should be optional. I can imagine where the same TLS configuration is shared between different Quarkus systems, so referring to it would be nice, but right now, I'm not seeing why we should update all properties to use a named TLS configuration

I deprecated OIDC-specific properties because of this sentence in ADR: https://github.com/quarkusio/quarkus/blob/860900eba562ac2bd82ff7bd342799ed15d5e0f7/adr/0004-using-the-tls-registry-for-clients.adoc?plain=1#L174

So, yes, there can be one extra property in case the TLS registry is only configured for let's say JWT credentials via keystore. On the other hand, OIDC-specific code in place only works for resources and JKS judging by the `io.quarkus.oidc.common.runtime.OidcCommonUtils#clientJwtKey`. While Quarkus TLS Registry also supports absolute paths and PKCS12.

TL;DR; if you want to keep existing OIDC-specific stuff in place, nothing forces us to deprecate it. I don't have strong opinion, honestly, these are little things. Let me know.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 1 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: That's correct @sberyozkin . But isn't this always true? If every extensions allows to configure extension specific properties, you will avoid one extra property because you don't need to add link between such extension and TLS registry. So was my thinking to deprecate it. But it is not important to do that.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: one more question, if we keep both, should I add new example as alternative? I think it might be bit confusing to have them alongside.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: meaning docs examples one next to another

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: hmm, I can add whole separate example like "or you can do this", so it should be an issue.

**Comment by sberyozkin** [comment]
Hi @michalvavrik, thanks,  let me get back to you on it, I've been thinking, can we have a convention, like a default oidc tls registry name, say `oidc`, so that that extra property is implied for simple setups ?

I'd also consider giving an example of the old approach and explain it may be eventually gone.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 1 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:200]: I would recommend using the tls-configuraiton-name to follow the ADR. OTherwise, your default configuration is going to also be used for the HTTP server, which is unlikely what you want (sharing between the HTTP server and OIDC)

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc [hunk 2]
```diff
@@ -266,21 +267,24 @@ The following example shows how you can configure `quarkus-oidc` to support `mTL
...
-quarkus.oidc.tls.verification=certificate-validation
+quarkus.oidc-client.tls.tls-configuration-name=oidc-client
+
+# configure hostname verification if necessary
+#quarkus.tls.oidc-client.hostname-verification-algorithm=NONE
 
 # Keystore configuration
-quarkus.oidc.tls.key-store-file=client-keystore.jks
-quarkus.oidc.tls.key-store-password=${key-store-password}
+quarkus.tls.oidc-client.key-store.jks.path=client-keystore.jks
```
Comment (cescoffier) [inline docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: not necessarily something to do in this PR, but can we move away from JKS, pretty please???

Seriously, this format is deprecated for a long time. Use p12 instead.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java [hunk 3]
```diff
@@ -51,18 +52,20 @@ static PolicyEnforcer createPolicyEnforcer(OidcTenantConfig oidcConfig,
...
-            adapterConfig.setTruststorePassword(oidcConfig.tls.trustStorePassword.orElse("password"));
-            if (OidcCommonConfig.Tls.Verification.CERTIFICATE_VALIDATION == oidcConfig.tls.verification
-                    .orElse(OidcCommonConfig.Tls.Verification.REQUIRED)) {
+        if (!tlsConfigSupport.useTlsRegistry()) {
+            boolean trustAll = oidcConfig.tls.getVerification().isPresent()
+                    ? oidcConfig.tls.getVerification().get() == OidcCommonConfig.Tls.Verification.NONE
+                    : tlsConfigSupport.isGlobalTrustAll();
+            if (trustAll) {
+                adapterConfig.setDisableTrustManager(true);
                 adapterConfig.setAllowAnyHostname(true);
```
Comment (cescoffier) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:61]: Is it the hostname verification? I guess so, but there are so many names referring to the same thing.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java [hunk 4]
```diff
@@ -86,7 +89,7 @@ static PolicyEnforcer createPolicyEnforcer(OidcTenantConfig oidcConfig,
                 .credentials(adapterConfig.getCredentials())
                 .bearerOnly(adapterConfig.isBearerOnly())
                 .enforcerConfig(enforcerConfig)
-                .httpClient(new HttpClientBuilder().build(adapterConfig))
+                .httpClient(new HttpClientBuilder().sslContext(tlsConfigSupport.getSslContext()).build(adapterConfig))
```
Comment (cescoffier) [inline extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:92]: What kind of HTTP client is it? 
If it's vert.x based, it might be better to use the dedicated Vert.x options. They support a lot more parameters than the SSL context.

**Review by cescoffier — CHANGES_REQUESTED** [review]
Looks good - I just asked a few questions.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:61]: @cescoffier I don't think you need to worry about this code, all I did is to put it behind flag `!useTlsRegistry` and I take global trust all flag from `tlsConfigSupport.isGlobalTrustAll()` instead of registry, but it's the same source of information. So this is the old behavior.

What matters is the code you commented below in the KeycloakPolicyEnforcerUtil.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java (hunk 4 above)
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:92]: It has to be `org.apache.http.client.HttpClient` because of what Keycloak PolicyEnforcer accepts, it's out of our hands. Therefore it's either SSL context or manually mapping from config https://github.com/quarkusio/quarkus/blob/0a21b51e652e523cadf7866f5963db169d72e66f/extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/HttpClientBuilder.java#L286

IMO we barely hear about KC AuthZ so I wonder how much is it used, but we could ask Pedro for Vert.X based client.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 2 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: I agree; the reason why I used JKS is that I wanted it one to one with the old docs. I'll update it when I push changes that @sberyozkin will suggest later (dropping deprecation or some convetion). Thank you

**Comment by michalvavrik** [comment]
> I'd also consider giving an example of the old approach and explain it may be eventually gone.

FYI @sberyozkin , when the old config is deprecated, it disappears from list of configuration properties, so it is either keeping the OIDC-specific properties not deprecated, or removing them from docs. Both does not make sense.

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:61]: but if you are interested, this original behavior did create SSL context that allowed all host names and `javax.net.ssl.X509TrustManager` that always trusted (NOOP)

File: extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java (hunk 4 above)
Comment (cescoffier) [reply extensions/keycloak-authorization/runtime/src/main/java/io/quarkus/keycloak/pep/runtime/KeycloakPolicyEnforcerUtil.java:92]: ah ok! Thanks!

**Review by cescoffier — APPROVED** [review]

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 2 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: @cescoffier @michalvavrik I don't really mind, I don't recall that the OIDC code insists somewhere on the JKS format

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: May be it was a copy and paste, and there is indeed some code in OidcCommonUtils that does it, not for this PR though

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: @michalvavrik I think you can just update that example to show what @cescoffier strongly recommends

File: docs/src/main/asciidoc/security-openid-connect-client-reference.adoc [hunk 5]
```diff
@@ -786,8 +786,9 @@ quarkus.oidc-client.credentials.jwt.key-file=privateKey.pem
 ----
 quarkus.oidc-client.auth-server-url=http://localhost:8180/auth/realms/quarkus/
 quarkus.oidc-client.client-id=quarkus-app
-quarkus.oidc-client.credentials.jwt.key-store-file=keystore.jks
-quarkus.oidc-client.credentials.jwt.key-store-password=mypassword
+quarkus.oidc-client.credentials.jwt.tls-configuration-name=oidc-client
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-client-reference.adoc:789]: @michalvavrik This is actually wrong, I should've clarified, sorry.

It is not about TLS at all, but about creating a signed JWT token from the keystore, so the key store configuration related to the `credentials.jwt` group must not be changed, please update, sorry again for not making it clear, I kind of referred to it in the issue, but obviously not clearly enough.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
+package io.quarkus.oidc.client;
```
Comment (sberyozkin) [inline extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: Unfortunately this test will have to go

File: extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcClientCommonConfig.java [hunk 7]
```diff
@@ -205,7 +205,8 @@ public static enum Source {
 
             /**
              * If provided, indicates that JWT is signed using a secret key.
-             * It is mutually exclusive with {@link #key}, {@link #keyFile} and {@link #keyStore} properties.
+             * It is mutually exclusive with {@link #key}, {@link #keyFile}, {@link #tlsConfigurationName} and {@link #keyStore}
```
Comment (sberyozkin) [inline extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcClientCommonConfig.java:208]: Credentials.Jwt should stay unchanged

**Review by sberyozkin — CHANGES_REQUESTED** [review]
Sorry Michal,  all the changes impacting Credentials.JWT should be reverted

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java (hunk 6 above)
Comment (sberyozkin) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: But updating `integration-tests/oidc-mtls` would be interesting, as there we have HTTP server MTLS, so testing it alongside the OIDC MTLS authentication with the tls registry config would be good

**Comment by michalvavrik** [comment]
> Sorry Michal, all the changes impacting Credentials.JWT should be reverted

Well, not that I mind, but I don't understand you. I won't answer individual comments because it makes sense to discuss this at one place. Here it is:

Why cannot user use one shared registry among `oidc.credentials.jwt` and `oidc.tls` and possibly others? We require user to specify key id and key id password. I think you are too quick to asking removing this feature why I can't see downside. Could you please suggest what is downside of keeping this extra option? I think there are only upsides, for example, extensions can provide their own TLS config, because TLS registry is extendable.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java (hunk 6 above)
Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: I thought `integration-tests/oidc` already is already testing mTLS between provider and Quarkus app. Maybe I read it wrong.

Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: Because `integration-tests/oidc-mtls` tests mTLS between app user requests, not provider and Quarkus. I will put in additional work if there is positive in there, I just tried to avoid adjustments in Keycloak resource side.

**Comment by michalvavrik** [comment]
> Sorry Michal, all the changes impacting Credentials.JWT should be reverted

@sberyozkin to quicken this discussion bit and cut down your time, you are basically saying it's called _TLS registry_ not _Keystore registry_ so we cannot use it for this purpose because here, we don't configure TLS, right? Yeah, I don't have answer for that, apart of that I don't care how is it called, it's positive feature.

Let me remove it and wait if there is ever user request for this option.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java (hunk 6 above)
Comment (michalvavrik) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: I'll remove this test.

File: docs/src/main/asciidoc/security-openid-connect-client-reference.adoc (hunk 5 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-client-reference.adoc:789]: There is no need to clarify, I was aware and included this because I disagree. Let me remove this.

File: extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java (hunk 6 above)
Comment (sberyozkin) [reply extensions/oidc-client/deployment/src/test/java/io/quarkus/oidc/client/OidcClientCredentialsJwtTlsRegistryTest.java:1]: @michalvavrik `integration-tests/oidc` does test it, you are right, I thought testing a situation where a TLS registry is used to configures both HTTP server and OIDC can be useful to verify, but I can definitely try it later, that is fine

File: docs/src/main/asciidoc/security-openid-connect-client-reference.adoc (hunk 5 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-openid-connect-client-reference.adoc:789]: Aware TLS != JWT credentials

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 2 above)
Comment (michalvavrik) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc:276]: FYI @sberyozkin I was referring to this "JKS" code. if this can run also PKCS12, I didn't try it so sorry for confusion. I think it's just JKS.

https://github.com/quarkusio/quarkus/blob/a486b5db0f1feec3088a8addb8002e91955e62ae/extensions/oidc-common/runtime/src/main/java/io/quarkus/oidc/common/runtime/OidcCommonUtils.java#L387

**Comment by sberyozkin** [comment]
Yeah, thanks @michalvavrik, things like trust all, etc, do not apply in this case. This signed token might travel over MTLS to OIDC but I'd say it would be either MTLS or just plain HTTP plus this token.
FYI, it is a `private_kkey_jwt` case covered at https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

**Comment by michalvavrik** [comment]
> Yeah, thanks @michalvavrik, things like trust all, etc, do not apply in this case. This signed token might travel over MTLS to OIDC but I'd say it would be either MTLS or just plain HTTP plus this token. FYI, it is a `private_kkey_jwt` case covered at https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication

Copy that, will push changes in couple of minutes.

File: integration-tests/oidc/src/main/resources/application.properties [hunk 8]
```diff
@@ -9,6 +9,18 @@ quarkus.oidc.tls.trust-store-file=client-truststore.jks
 quarkus.oidc.tls.trust-store-password=password
 quarkus.oidc.tls.key-store-file=client-keystore.jks
 quarkus.oidc.tls.key-store-password=password
+
+%tls-registry.quarkus.oidc.tls.tls-configuration-name=oidc-tls
```
Comment (sberyozkin) [inline integration-tests/oidc/src/main/resources/application.properties:13]: @michalvavrik Can you clarify please this setup ? Why to declare empty legacy properties ?

Comment (michalvavrik) [reply integration-tests/oidc/src/main/resources/application.properties:13]: Because Quarkus config profile `tls-registry` has priority, but properties set without this profile are still effective. Have a look at lines 9, 10, 11. It is not a trick, when I set `=` I am leveling the ground to have same state, as when you don't set it.

Comment (michalvavrik) [reply integration-tests/oidc/src/main/resources/application.properties:13]: Maybe I should add additional validation to make old and new properties exclusive. My thinking was that users won't see the old properties and that nothing bad happens if they configure both.

File: docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc (hunk 2 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-code-flow-authentication.adoc]: Thanks @michalvavrik, yeah, I'll open a bug request to get rid of hard-coding to JKS

**Review by sberyozkin — APPROVED** [review]
